### PR TITLE
ENH: Use Masked array repr convention in NDData

### PR DESCRIPTION
**Pull request by bmorris3**
### Description

`NDData` objects currently represent themselves via `__repr__` by their `data` attribute, while ignoring their masks. For example, if you do:
```python
>>> import numpy as np
>>> import astropy.units as u
>>> from astropy.nddata import NDDataArray, StdDevUncertainty, NDData

>>> data = np.array([1,2,3,4])
>>> mask = data > 2
>>> unit = 'erg / s'
>>> uncertainty = StdDevUncertainty(np.sqrt(data))
>>> ndd = NDData(data, mask=mask, unit=unit, uncertainty=uncertainty)
>>> ndd
NDData([1, 2, 3, 4], unit='erg / s')
```
...the output does not reflect that (a) there is a mask, (b) the mask is not all False.

This PR ~uses the numpy utility [`numpy.ma.MaskedArray._insert_masked_print`](https://github.com/numpy/numpy/blob/b175756dd358f659a0249461934cae37bae534f3/numpy/ma/core.py#L3955) within `NDData.__repr__`,~ `astropy.utils.Masked` so that the previous example will now print:
```python
>>> ndd
NDData([1, 2, --, --], unit='erg / s')
```
This simply makes the mask visible in the repr – the underlying `data` attribute isn't touched by this PR.

There are changes to docs related to `NDData` in this PR which ensure the string outputs match after this formatting tweak is applied.

### CODE REVIEW

File: astropy/nddata/nddata.py [hunk 1]
```diff
@@ -263,7 +263,8 @@ def __str__(self):
 
     def __repr__(self):
         prefix = self.__class__.__name__ + "("
-        data = np.array2string(self.data, separator=", ", prefix=prefix)
+        ma = np.ma.masked_array(self.data, self.mask)
+        data = np.array2string(ma._insert_masked_print(), separator=", ", prefix=prefix)
```
[the remark this answers is not in the document]
Comment (bmorris3) [reply astropy/nddata/nddata.py:267]: We certainly don't have to, but it was economical to use the existing code, and I don't foresee the output from this method changing much. 🤷🏻

[the remark this answers is not in the document]
Comment (bmorris3) [reply astropy/nddata/nddata.py:267]: Sounds good, on it.

File: astropy/nddata/nddata.py [hunk 2]
```diff
@@ -263,7 +263,8 @@ def __str__(self):
 
     def __repr__(self):
```
Comment (Cadair) [inline astropy/nddata/nddata.py:264]: This is sort of un-related, but the `__repr__` using numpy functions to print the *values* in the array, interacts really badly with non-numpy arrays like Dask.

I would rather see all this code only run if the data and/or masks are only subo;asses of numpy arrays.

Comment (bmorris3) [reply astropy/nddata/nddata.py:265]: @Cadair good idea. Somewhat miraculously, if I do: 
```python
import numpy as np
import dask.array as da
from astropy.nddata import NDData

NDData(da.array(np.arange(5)), mask=[False, True, False, True, False])
```
I get 
```
NDData([0 — 2 — 4])
```
Is that ok for you?

File: astropy/nddata/nddata.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmorris3) [reply astropy/nddata/nddata.py]: Implemented in [aa5c602](https://github.com/astropy/astropy/pull/14438/commits/aa5c60294588219d7932fb42a181127351ffb548).

File: astropy/nddata/nddata.py (hunk 2 above)
Comment (bmorris3) [reply astropy/nddata/nddata.py:265]: I don't fully understand what's happening underneath – so it's entirely possible the numpy calls are converting the dask array to a numpy array, and this single line would melt my MacBook if that were a typical DKIST dask array.

Comment (mhvk) [reply astropy/nddata/nddata.py:265]: I fear your are right that as it is, you might run into trouble... But with `Masked`, it should be fairly easy to change, as it overrides `array2string` & co. For large arrays, only the edges are shown, so the conversion to array would have to be moved after those have been selected.

p.s. If you are interested, the override is around https://github.com/astropy/astropy/blob/2aef9dc327fbc9fd29fb2ced2be77eb7f3f625ec/astropy/utils/masked/function_helpers.py#L975

Comment (Cadair) [reply astropy/nddata/nddata.py:265]: I think it definitely will load the dask array, and last I checked it computed the whole dask array, not just the bits it ended up displaying.

I would suggest we include `.data.__repr__` in the `.NDData.__repr__` because then it's agnostic to the array type?

Comment (bmorris3) [reply astropy/nddata/nddata.py:265]: I'm happy with that, but it means that the concise masking representation is out of scope. 

Numpy gives us one hint on how to dodge this problem in their masked array repr:
```python
>>> ma
masked_array(
  data=[[--, 2, 3],
        [2, 3, 4],
        [3, --, 5]],
  mask=[[ True, False, False],
        [False, False, False],
        [False,  True, False]],
  fill_value=999999)
```
We could just show each NDData attribute separately. 

Pros: 
* we can rely on each attribute's repr and remain agnostic about `type(nddata.data)`
* it's pretty clear what's in the object
* it's easy to implement

Cons: 
* it's less pretty than either @mhvk's or numpy's implementation, which both obscure masked elements.

Here's an implementation,
```python
prefix = nddata.__class__.__name__ + "("
contents = []
for attr in ['data', 'mask', 'uncertainty', 'unit']:
    attr_data = getattr(nddata, attr)
    if attr_data is not None:
        attr_prefix = f'\n  {attr}='
        attr_repr = repr(attr_data)
        attr_repr = attr_repr.replace("\n", f'\n{" " * (len(attr_prefix) - 1)}')
        contents.append(attr_prefix + attr_repr)
return prefix + ','.join(contents) + ")"
```
and using this implementation we can make an NDData from dask
```python
import numpy as np
import dask.array as da
from astropy.nddata import NDData

nddata = NDData(
    da.array(np.arange(5_000_000).reshape((250000, 20))), 
    mask=np.random.randint(0, 2, size=5_000_000).astype(bool).reshape((250000, 20))
)
```
which prints like this:
```
NDData(
  data=dask.array<array, shape=(250000, 20), dtype=int64, chunksize=(250000, 20), chunktype=numpy.ndarray>,
  mask=array([[ True,  True, False, ..., False, False, False],
              [False, False, False, ...,  True, False,  True],
              [ True,  True, False, ..., False,  True,  True],
              ...,
              [False,  True, False, ...,  True, False, False],
              [False, False, False, ...,  True, False, False],
              [False,  True, False, ..., False,  True, False]]))
```

Thoughts, @Cadair, @mhvk?

Comment (mhvk) [reply astropy/nddata/nddata.py:265]: My sense would be to use `Masked` if `.data` is an `ndarray` subclass and otherwise just `.data.__repr__` as you suggest. Hopefully, we can adjust `Masked` to work on `dask` arrays itself - unlikely `Quantity`, it is meant to be somewhat agnostic to what is being masked.

But I can also see the case for keeping it simple and just rely on the `__repr__` of each part.

Indeed, for a `__repr__` it is generally good to have format that is close to what is needed to initialize. I guess perhaps the bigger question is whether one wants a `__str__` as well, where one would (I think) to not show the mask explicitly (and where `Masked(nd.data << nd.unit, mask=nd.mask)` would do OK...)

File: docs/changes/nddata/14438.feature.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
+A more flexible and/or compact string representation is now available for
```
Comment (mwcraig) [inline docs/changes/nddata/14438.feature.rst:1]:
```suggestion
A more flexible representation is now available for
```

I would drop the "and/or" and remove "string" because we have changed the `__repr__` but not `__str__`.

**Review by mwcraig — COMMENTED** [review]
Looking good, just one question  and one suggestion.

File: astropy/nddata/mixins/ndslicing.py [hunk 4]
```diff
@@ -30,7 +30,7 @@ class NDSlicingMixin:
 
         >>> nd = NDDataSliceable([1,2,3,4,5])
         >>> nd[1:3]
-        NDDataSliceable([2, 3])
+        NDDataSliceable([2 3])
```
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/nddata/mixins/ndslicing.py:33]: That's actually a bit confusing, since the `__repr__` of `Masked` does have `,`; it follows what `ndarary` (and `np.ma.MaskedArray`) does:
```
In [10]: str(Masked(np.array([1, 2.])))
Out[10]: '[1. 2.]'

In [11]: repr(Masked(np.array([1, 2.])))
Out[11]: 'MaskedNDArray([1., 2.])'

In [12]: str(np.array([1, 2.]))
Out[12]: '[1. 2.]'

In [13]: repr(np.array([1, 2.]))
Out[13]: 'array([1., 2.])'
```

I agree that here it should have`,` too -- maybe have a look at what I did for `MaskedNDArray`?

**Comment by pllim** [comment]
Is this ready for a final review before feature freeze?

**Review by mwcraig — CHANGES_REQUESTED** [review]
Sorry for the long delay here, @bmorris3. 

I think this is ready to go except for putting the comma back in the `repr` to match astropy `Masked` and `numpy.ma.MaskedArray`.

**Comment by bmorris3** [comment]
I've fixed the missing-comma issue that we [last discussed above](https://github.com/astropy/astropy/pull/14438#discussion_r1136035926). Thanks for the tips, @mwcraig & @mhvk!

**Review by mwcraig — APPROVED** [review]

**Comment by mwcraig** [comment]
Thanks @bmorris3!

File: docs/whatsnew/5.3.rst
Comment (pllim) [inline docs/whatsnew/5.3.rst:1]: Why this change?

Comment (bmorris3) [reply docs/whatsnew/5.3.rst:1]: Blegh that's a rebase issue, apologies. Should I make a fix PR?

Comment (bmorris3) [reply docs/whatsnew/5.3.rst:1]: Blegh that's a rebase issue. Fixing!

Comment (bmorris3) [reply docs/whatsnew/5.3.rst:1]: Thanks for catching it!

**Review by pllim — CHANGES_REQUESTED** [review]
Can we wait till https://github.com/astropy/astropy/pull/15177 is merged and then rebase this again? I am afraid this might trip something in numpy-dev because of all the repr stuff.

**Comment by bmorris3** [comment]
@pllim Sure thing! Could you ping me when that happens?

**Comment by pllim** [comment]
That other PR is merged. Please rebase, thanks!

File: astropy/nddata/nddata.py (hunk 2 above)
Comment (pllim) [reply astropy/nddata/nddata.py:308]: So, is this resolved or something still needs to be done here?

File: docs/nddata/subclassing.rst [hunk 5]
```diff
@@ -156,16 +156,16 @@ on the class or the instance::
     >>> ndd = NDDataPower([1,2,3])
 
     >>> # using it on the instance with one operand
-    >>> ndd.pow(3)
+    >>> ndd.pow(3)  # doctest: +SKIP
```
Comment (pllim) [inline docs/nddata/subclassing.rst:159]: Why are these three doctest skipped?

Comment (bmorris3) [reply docs/nddata/subclassing.rst:159]: It turns out the numpy repr formatting also prints the dtype, but only on Windows. The skip avoids an error on the windows test.

Comment (pllim) [reply docs/nddata/subclassing.rst:159]: Is there no way to use ellipsis? Do you still have the error log on Windows? Seems excessive to skip a valid doctest just because of one OS/arch.

p.s. Maybe we have to fast track dropping doctest on certain platforms but that is a discussion for another day.

Comment (bmorris3) [reply docs/nddata/subclassing.rst:159]: Here's the failure: https://github.com/astropy/astropy/actions/runs/5859004595/job/15884145980#step:10:1734

Are you suggesting something like this?
```diff
-     >>> ndd.pow(3)  # doctest: +SKIP
-    NDDataPower([ 1,  8, 27])

+    >>> ndd.pow(3)  # doctest: +ELLIPSIS
+    NDDataPower([ 1,  8, 27]...)
```
Personally I think ellipsis is more confusing.

Comment (pllim) [reply docs/nddata/subclassing.rst:159]: https://github.com/scientific-python/pytest-doctestplus#doctest-dependencies allows conditional skipping though I never tried it with platform check. Maybe we can do that. If not, please open a follow-up issue to un-skip this later because I feel like this should not be skipped if possible.

Also I am going to cc @mhvk here since this might add to the things he has to handle over at #15065

Comment (bmorris3) [reply docs/nddata/subclassing.rst:159]: I'm happy to open an issue. It'd be great to get it working correctly everywhere, but it seems like the easiest way to do that now is to add conditional repr formatting in astropy to clean up the platform differences.

Comment (bmorris3) [reply docs/nddata/subclassing.rst:159]: Think I got it in [0e57b49](https://github.com/astropy/astropy/pull/14438/commits/0e57b49c3439129221c35a6a6037b6ba21c908ac)!

Comment (pllim) [inline docs/nddata/subclassing.rst:159]: Missed one?

```suggestion
    >>> ndd.pow(3)  # doctest: +ELLIPSIS
```

Comment (bmorris3) [reply docs/nddata/subclassing.rst]: Thanks!

**Comment by pllim** [comment]
You might need another rebase to pick up numpy dev compat fixes.

File: docs/nddata/subclassing.rst (hunk 5 above)
Comment (pllim) [inline docs/nddata/subclassing.rst:159]: What about this one?
```suggestion
    >>> ndd.pow(3)  # doctest: +ELLIPSIS
```

**Review by pllim — COMMENTED** [review]
Need yet another rebase to pick up RTD fix, sorry.

File: docs/nddata/subclassing.rst (hunk 5 above)
Comment (bmorris3) [reply docs/nddata/subclassing.rst]: Sorry, I did the last rebase without pulling the committed change from your suggestion. I won't do that this time.

File: docs/nddata/subclassing.rst [hunk 6]
```diff
@@ -156,16 +156,16 @@ on the class or the instance::
     >>> ndd = NDDataPower([1,2,3])
 
     >>> # using it on the instance with one operand
-    >>> ndd.pow(3)
+    >>> ndd.pow(3)  # doctest: +ELLIPSIS
     NDDataPower([ 1,  8, 27])
```
Comment (pllim) [inline docs/nddata/subclassing.rst:160]:
```suggestion
    NDDataPower([ 1,  8, 27]...)
```

**Comment by pllim** [comment]
I don't understand why Parallel job is failing and the log makes no sense but it is not related.

**Comment by pllim** [comment]
Huh weird, some jobs seem to be stuck.

File: astropy/nddata/nddata.py [hunk 7]
```diff
@@ -307,10 +307,38 @@ def __str__(self):
...
+            # and re-pad multi-line array outputs
+            after_first_paren = data_repr.index("(") + 1
+            old_prefix_spaces = " " * after_first_paren
+            new_prefix_spaces = " " * len(prefix)
+            data_repr = data_repr[after_first_paren:-1].replace(
+                old_prefix_spaces, new_prefix_spaces
+            )
+            unit = f", unit='{self.unit}'" if self.unit is not None else ""
+            return f"{prefix}{data_repr}{unit})"
+        else:
```
Comment (pllim) [inline astropy/nddata/nddata.py:328]: Codecov says this "else" is not covered.

Comment (bmorris3) [reply astropy/nddata/nddata.py:328]: @pllim This `else` covers cases where `data` is a dask array, see discussion in https://github.com/astropy/astropy/pull/14438#discussion_r1117116482. I'm not sure if that can/should be covered in our tests? cc @Cadair

Comment (pllim) [reply astropy/nddata/nddata.py:328]: We do have dask here:

https://github.com/astropy/astropy/blob/7bc600506dad3c1ce697a68fcbbc395256a438a4/setup.cfg#L78

Comment (bmorris3) [reply astropy/nddata/nddata.py:337]: Dask test coverage introduced 👌🏻

File: astropy/nddata/nddata.py [hunk 8]
```diff
@@ -307,10 +312,42 @@ def __str__(self):
...
+            # relying on the builtin repr for objects other than ndarrays
+            # allows for compatibility with e.g. dask:
+            contents = []
+            for attr in ["data", "mask", "uncertainty", "unit"]:
```
Comment (pllim) [inline astropy/nddata/nddata.py:341]: Nitpick. I don't think there is a difference but I prefer tuple in principle. Take it or leave it.

```suggestion
            for attr in ("data", "mask", "uncertainty", "unit"):
```

File: astropy/nddata/tests/test_nddata.py [hunk 9]
```diff
@@ -22,6 +22,8 @@
 
 from .test_nduncertainty import FakeUncertainty
 
+da = pytest.importorskip("dask.array")
```
[the remark this answers is not in the document]
Comment (bmorris3) [reply astropy/nddata/tests/test_nddata.py:25]: Ah I only want the later. Should I put this line in the test that needs it?

File: astropy/nddata/nddata.py (hunk 2 above)
Comment (pllim) [reply astropy/nddata/nddata.py:313]: Is this resolved?

**Review by pllim — COMMENTED** [review]
Tests LGTM now. Thanks!

@mhvk or @Cadair , is this good to go in? (Please use "Squash an Merge" button!)

**Comment by pllim** [comment]
@bmorris3 , alas, there is a conflict now. There is no more `setup.cfg` so please adjust your patch accordingly. I apologize for any inconvenience caused. Thanks!

**Comment by bmorris3** [comment]
@pllim Rebased, thanks!

**Comment by pllim** [comment]
I will wait a couple more hours to grab the new scipy dev wheel and see if devdeps would pass. 🤞

**Comment by pllim** [comment]
devdeps failure is now caused by IERS instead of scipy, but still unrelated.

Never heard back from Stuart, so let's just get this in.
